src/rand.c

```diff
@@ -0,0 +1,81 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include "rand.h"
+#include "tinymt64.h"
+
+// TODO Add support for cryptographically secure PRNG
+// TODO Add Windows CNG
+
+static tinymt64_t fast_rand;
+
+#if HAVE_GETRANDOM_SYSCALL
+// -------------------
+// Modern linux kernel
+// -------------------
+// Use getrandom syscall which has some advantages over /dev/urandom
+
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <linux/random.h>
+
+void jq_rand_init() {
+  uint64_t seed;
+  if (syscall(SYS_getrandom, &seed, sizeof(seed), 0) != sizeof(seed)) {
+    fprintf(stderr, "error: could not get data from random pool.\n");
+    exit(2);
+  }
+  tinymt64_init(&fast_rand, seed);
+}
+
+#elif HAVE__DEV_URANDOM
+// -------------------
+// Unix machine
+// -------------------
+// Use /dev/urandom as fallback
+
+void jq_rand_init() {
+  uint64_t seed;
+  FILE *devur = fopen("/dev/urandom", "r");
+  if (fread(&seed, sizeof(seed), 1, devur) != 1) {
```

Comment:
Need to check that `devur` isn't NULL...  That should raise an error.

Ideally we could #ifdef out on `WIN32` so that on Windows these builtins elicit a compile-time error.

We should also have a notion of unsupported builtins so that a compile-time error about them can be more user-friendly.

And, lastly, WIN32 support, using CNG, would be nice.

Lastly, `getentropy(2)` or `getrandom(2)` might be better than reading from `/dev/urandom` where those are supported.

To be clear, I'm not asking for Windows support though, or for support for `getentropy(2)`/`getrandom(2)`.

Reply:
This does actually check at compile time for the existence of the "getrandom" syscall, and the "/dev/urandom" file. It's still possible for "dev/urandom" to be null, so you're right, it should be checked.